src/rt/object.rs

```diff
@@ -53,106 +51,106 @@ pub(crate) enum Action {
 }
 
 impl Object {
-    pub(super) fn atomic_mut(&mut self) -> Option<&mut atomic::State> {
-        match self.kind {
-            Kind::Atomic(ref mut v) => Some(v),
+    pub(super) fn atomic_mut(self, store: &mut Store) -> Option<&mut atomic::State> {
+        match &mut store.entries[self.id] {
+            Entry::Atomic(v) => Some(v),
             _ => None,
         }
     }
 
-    pub(crate) fn mutex() -> Object {
-        Object {
-            kind: Kind::Mutex(None),
+    pub(super) fn condvar_mut(self, store: &mut Store) -> Option<&mut condvar::State> {
+        match &mut store.entries[self.id] {
+            Entry::Condvar(v) => Some(v),
+            _ => None,
         }
     }
 
-    pub(crate) fn condvar() -> Object {
-        Object {
-            kind: Kind::Condvar(None),
+    pub(super) fn mutex_mut(self, store: &mut Store) -> Option<&mut mutex::State> {
+        match &mut store.entries[self.id] {
+            Entry::Mutex(v) => Some(v),
+            _ => None,
         }
     }
 
-    pub(crate) fn thread() -> Object {
-        Object {
-            kind: Kind::Thread(None),
+    pub(super) fn notify_mut(self, store: &mut Store) -> Option<&mut notify::State> {
+        match &mut store.entries[self.id] {
+            Entry::Notify(v) => Some(v),
+            _ => None,
         }
     }
 }
 
-impl Set {
-    pub(crate) fn new() -> Set {
-        Set { objects: vec![] }
+impl Store {
+    /// Create a new, empty, object store
+    pub(crate) fn new() -> Store {
+        Store { entries: vec![] }
+    }
+
+    /// Insert a new atomic object into the store.
+    pub(super) fn insert_atomic(&mut self, state: atomic::State) -> Object {
```

Comment:
It seems like there's some code here (e.g. `insert_atomic`, `insert_mutex`, insert_condvar`, ...) that could potentially be de-duplicated using traits, maybe
```rust
pub(super) fn insert(&mut self, impl Into<Entry>) -> Object
```
or something?

Not a blocker though; I'm not sure if this actually matters.

Reply:
I think a trait strategy could reduce some boilerplate. I opted to punt for this PR.